presto-native-execution/presto_cpp/main/operators/StorageBasedLocalShuffle.cpp

```diff
@@ -0,0 +1,128 @@
+
+namespace {
+inline std::string
+createShuffleFilename(std::string rootPath, int32_t partition, int fileIndex) {
+  return rootPath + "/" + std::to_string(partition) + "_" +
+      std::to_string(fileIndex) + ".bin";
+}
+}; // namespace
+
+namespace facebook::presto::operators {
+
+std::unique_ptr<velox::WriteFile> StorageBasedLocalShuffle::getNextOutputFile(
+    int32_t partition) {
+  auto fileCount = getPartitionFilesCount(partition);
+  std::string filename = createShuffleFilename(rootPath_, partition, fileCount);
+  return fileSystem_->openFileForWrite(filename);
+}
+
+bool StorageBasedLocalShuffle::fileExists(const std::string& filename) const {
+  if (fileSystem_->name() == "Local FS") {
+    return std::filesystem::exists(filename);
+  }
+  VELOX_FAIL("Only supports local filesystem.");
+}
+
+int StorageBasedLocalShuffle::getPartitionFilesCount(int32_t partition) const {
+  auto fileCount = 0;
+  bool fileExist = false;
+  do {
+    std::string filename =
+        createShuffleFilename(rootPath_, partition, fileCount);
+    fileExist = fileExists(filename);
+    if (!fileExist) {
+      break;
+    }
+    fileCount++;
+  } while (fileExist);
+  return fileCount;
+}
+
+void StorageBasedLocalShuffle::storePartitionBlock(int32_t partition) {
+  auto& buffer = inProgressPartitions_[partition];
+  buffer->setSize(inProgressSizes_[partition]);
+  auto file = std::move(getNextOutputFile(partition));
+  file->append(std::string_view(buffer->asMutable<char>(), buffer->size()));
+  file->close();
+  inProgressPartitions_[partition].reset();
+}
+
+void StorageBasedLocalShuffle::collect(
+    int32_t partition,
+    std::string_view data) {
+  auto& buffer = inProgressPartitions_[partition];
+
+  // Check if there is enough space in the buffer.
+  if (buffer &&
+      inProgressSizes_[partition] + data.size() + sizeof(size_t) >=
+          maxBytesPerPartition_) {
+    storePartitionBlock(partition);
+  }
+
+  // Allocate buffer if needed.
+  if (!buffer) {
+    buffer = AlignedBuffer::allocate<char>(maxBytesPerPartition_, pool_);
+    inProgressSizes_[partition] = 0;
+  }
+
+  // Copy data.
+  auto rawBuffer = buffer->asMutable<char>();
+  auto offset = inProgressSizes_[partition];
+
+  *(size_t*)(rawBuffer + offset) = data.size();
+
+  offset += sizeof(size_t);
+  memcpy(rawBuffer + offset, data.data(), data.size());
+
+  inProgressSizes_[partition] += sizeof(size_t) + data.size();
+}
+
+void StorageBasedLocalShuffle::noMoreData(bool success) {
+  VELOX_CHECK(success, "Unexpected error.")
+  for (auto i = 0; i < numPartitions_; ++i) {
+    if (inProgressSizes_[i] > 0) {
+      storePartitionBlock(i);
+    }
+  }
+}
+
+bool StorageBasedLocalShuffle::hasNext(int32_t partition) const {
+  if (readPartitionsFileIndex_[partition] < getPartitionFilesCount(partition)) {
+    return true;
+  }
+  return false;
+}
+
+BufferPtr StorageBasedLocalShuffle::next(int32_t partition, bool success) {
+  VELOX_CHECK(success, "Unexpected error.")
```

Comment:
Can we handle this case?

Reply:
I think so. The best we can do is to reset the index of the files to be read for that partition to 0.